certora/specs/GovernorFunctions.spec

```diff
@@ -0,0 +1,277 @@
+import "helpers.spec"
+import "Governor.helpers.spec"
+import "GovernorInvariants.spec"
+
+use invariant proposalStateConsistency
+use invariant queuedImplyDeadlineOver
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: propose effect and liveness. Includes "no double proposition"                                                 │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule propose_liveness(uint256 pId, env e) {
+    require nonpayable(e);
+    require clockSanity(e);
+
+    uint8 stateBefore = state(e, pId);
+
+    address[] targets; uint256[] values; bytes[] calldatas; string descr;
+    require pId == hashProposal(targets, values, calldatas, descr);
+    //require validString(descr);
+
+    propose@withrevert(e, targets, values, calldatas, descr);
+
+    // liveness & double proposal
+    assert !lastReverted <=> (
+        stateBefore == UNSET() &&
+        validProposal(targets, values, calldatas)
+    );
+}
+
+rule propose_effect(uint256 pId, env e) {
+    address[] targets; uint256[] values; bytes[] calldatas; string descr;
+    require pId == hashProposal(targets, values, calldatas, descr);
+
+    propose(e, targets, values, calldatas, descr);
+
+    // effect
+    assert state(e, pId)         == PENDING();
+    assert proposalProposer(pId) == e.msg.sender;
+    assert proposalSnapshot(pId) == clock(e) + votingDelay();
+    assert proposalDeadline(pId) == clock(e) + votingDelay() + votingPeriod();
+}
+
+rule propose_sideeffect(uint256 pId, env e, uint256 otherId) {
+    uint8   otherStateBefore = state(e, otherId);
+    uint256 otherVoteStart   = proposalSnapshot(otherId);
+    uint256 otherVoteEnd     = proposalDeadline(otherId);
+    address otherProposer    = proposalProposer(otherId);
+
+    address[] targets; uint256[] values; bytes[] calldatas; string descr;
+    require pId == hashProposal(targets, values, calldatas, descr);
+
+    propose(e, targets, values, calldatas, descr);
+
+    // no side-effect
+    assert state(e, otherId)         != otherStateBefore => otherId == pId;
+    assert proposalSnapshot(otherId) == otherVoteStart;
+    assert proposalDeadline(otherId) == otherVoteEnd;
+    assert proposalProposer(otherId) == otherProposer;
```

Comment:
Slightly confused, why do these work even for `otherId == pId`?

Reply:
In this rule we do `state(e, otherId)`. This means that `otherId` has a valid state, which means its been proposed (because state reverts if the proposal doesn't exist).

So we know `otherId` exists, and as been propose. 

`propose_liveness` is supposed (if it was working) to prove that `propose(...)` reverts if the proposal already exist. Since this rule does call propose without `@withrevert`, we know the propose worked, which implies `otherId == pId`.

So technically, we (expect to) know for sure that `otherId == pId`.

Reply:
I think these (failing) rules should be rewritten using `safeState`